### ISSUE TITLE
Don't allow opening piano roll with invalid pattern

### ISSUE THREAD

**Issue by allejok96** [LMMS/lmms#6234]
User: I opened the piano, but I cannot write notes
Solution: tell user he is doing it wrong
Problem: user is clearly illiterate, and this is bad design*

Developer: piano-roll crashes when pattern is `nullptr`
Solution: add `hasValidPattern()`
Problem: now there are safety checks everywhere

What if we simply prevented piano roll window from opening when there is no valid pattern?

### Solution

1. Disable/gray out piano roll button when there is no valid pattern.

The way the main window toolbar and view menu are set up currently makes things harder for us. It would be easier if we could use a common `QAction` for toggling piano roll that simply could be disabled.

2. Close the piano-roll window when pattern is deleted.

`MainWindow::toggleWindow` can be used to force open windows, but not to close them. This whole implementation could need some cleaning up. There is even a [bugfix](https://github.com/LMMS/lmms/commit/cc1cdfa664e7fd548917ae4d9f879770015b34db) in there, but since some editors code doesn't rely on this function and decide to implement it themselves, the bug remains.

### * Opening a useless window is bad

Most modern software does not open a window if you cannot do anything in it. E.g. a text editor always has a document on the screen. In some cases the main window will be blank, but in those cases there is a Open button or something in the toolbar. There are a few programs out there that uses toolbox-like windows that are unusable until the correct action is taken in the main window (image editors etc), but it is fairly obvious that those small windows are helpers to the main window where 95% of the work will happen.

This is not the case for piano-roll. It actually opens a window that remains unusable until the user does the correct thing *in another window*. The design essentially leads the user into a dead end.

**Comment by musikBear** [comment]
You are referring to this:
![image](https://user-images.githubusercontent.com/6662066/145683454-8a98011b-394f-4731-a970-8048168cc03c.png)
right?
I have always disliked this window. It also has a text that confuse a newcomer more than help him.
The text implies that 'something' can be done in this situation, "open a pattern" -But it doers not explain where or what a 'pattern' may be (now called **clip** -so an edit is needed anywitch). 
But if nothing actually opens when the user clicks on 
![image](https://user-images.githubusercontent.com/6662066/145683628-1b36b9b8-2db4-42ce-9082-52262e3265f8.png)
-Then that would be confusing too.
A method that would highlight an empty clip in song-editor, 
![image](https://user-images.githubusercontent.com/6662066/145683842-c8a8d6fb-2396-408b-9f43-88db99583c37.png)
could point the user to the correct solution, but that would 'clash' with our UI in general.

**Comment by qnebra** [comment]
I think this message is pretty vague, maybe it should be something like "Please open a pattern in Song Editor by double-clicking on it"

**Comment by musikBear** [comment]
> "Please open a pattern in Song Editor by double-clicking on it"
Yes or even more Pedagogic like:
"You have opened an un-assigned piano-roll! Please close this window, and open a assigned-pattern in *Song Editor* by double-clicking on a beat"
There could even be a reference to gitBook

**Comment by allejok96** [comment]
No, stop right there. This issue is about removing the need for that text entirely.

>  if nothing actually opens when the user clicks on [piano button] then that would be confusing too

@musikBear, there are a few places in LMMS where options are disabled (gray) and does not let the user click them:

-"Clear ghost notes" when there is no ghost
- "Set as ghost" when there are no notes
- The import/export pattern button in piano roll, when there is no pattern

Most users will understand when a button is disabled, and they will look for other ways. If we stop a new user from opening a useless piano window, we are *forcing* him to experiment with the song editor. Eventually he will create a clip and open it in piano-roll, without reading a gitbook manual.

The next problem is that the default LMMS theme makes it hard to see a difference between buttons that are disabled and buttons that can be clicked.

**Comment by qnebra** [comment]
Hmm, doesn't LMMS had tooltip system? Like when user hover mouse cursor over patterns in Song Editor it shows something like "Double-click to open pattern in corresponding editor"

**Comment by RiedleroD** [comment]
Too many tooltips might clutter the UI - that being said, adding a tooltip to the _disabled_ piano roll _button_ with "Open a pattern to enable the piano roll" (or smth similar) might be extra helpful.

**Comment by Monospace-V** [comment]
There was a tooltip system, when hovering. It's gone now, not sure what happened to it.

**Comment by qnebra** [comment]
It was removed few years ago. LMMS still had tooltips, smaller and in bottom left corner, mainly used to show messages like "Loading plugin, please wait while loading the VST plugin..."

**Comment by allejok96** [comment]
Off topic: I would love to see is a quick and professional looking tutorial when opening lmms for the first time. Similar to how Android apps do it, you know when they highlight some button and explain what it's for... Show users the essentials about clips, piano, and how to do automation so the CNA start playing around. This could catch some users that download lmms just to try it out and see if they like it.

**Comment by musikBear** [comment]
> Tooltips

We used to have a feature "What-is-this", where a question-mark could be dragged into the UI, and a text, that had the nature of a in-program manual, would appear. That feature was deemed unpractical, because it needed to be edited with each new change. I still have the complete library of the WITs, though.

> Stop user

Greying out is good, but i fear it is a bit difficult to implement, since it is possible to open empty patterns, even in projects with functional clips. The logic for "when-to" grey out is tricky. Most likely the reason that it is not already done(?)

**Comment by reinierpost** [comment]
I'm sorry guys, I just installed LMMS and even after reading this I have no clue what to do.

After googling, I found someone who explained where to double-click, I do that and the Beat+Bassline Editor opens up, but the message doesn't go away. The connection between these things is completely unclear to me. Presumably a pattern is something the Beat+Bassline editor provides, but I don't want beats or basslines, I want to record my MIDI keyboard.

Usability-wise, this is a nightmare.

I"m uninstalling this software because I can't figure out how to use it.

You are losing users without any good reason. This issue has existed for 10 years!

**Comment by qnebra** [comment]
Use "Song Editor" instead, I don't even understand why many tutorials and so on had this insanely bad take that LMMS workflow is nearly the same as FL Studio workflow. It is not.

1. Open LMMS
2. Maximise "Song Editor", I personally dislike "Pattern Editor" and avoid this thing as much as I can
3. Drag and drop instruments/samples from Sidebar to Song Editor, by default it is on left
4. And there you can double-click on pattern and it works.

And additional part about MIDI recording:
5. Made sure you connected MIDI keyboard and it is running before opening LMMS
6. Let assume you had instrument track in "Song Editor"
7. Find gear menu, click it, go to MIDI, Input and select your device
![image](https://github.com/LMMS/lmms/assets/11295229/d259ddb4-9024-4578-b745-95780e2a2179)
8. Then double-click on pattern
9. Press "Record notes from midi device" button

**Comment by musikBear** [comment]
> I want to record my MIDI keyboard.

Well if you un-installed then this is futile, but you do need to read the manual for software imo...
Any witch
left click in song-editor on an empty spot of your choice -eg where you want to place notes..
The spot turns grey
Right click it
Select Open-in-piano-roll
Done..
Now you can 'record' et catch notes in piano-roll as you play them with either of these transport-buttons
https://docs.lmms.io/user-manual/3-navigating-lmms/piano-roll#id-3.8.1.1-playback-and-record
This is help request, not a bug, and does not belong here.
